### PR TITLE
Added project files.

### PR DESCRIPTION
**Pull request by Larisa-Antsifrova**
Done:
- Added files from the case task.

Remember:
- Deadline 6.08.2021 23:59.

Happy reviewing :)

### CODE REVIEW

File: repositories/users-repository.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      id,
+      name,
+      email,
+      password,
+    };
+
+    const allUsers = await getAllUsers(usersPath);
+    allUsers.push(newUser);
+
+    await fs.writeFile(usersPath, JSON.stringify(allUsers, null, 2));
```
Comment (MazanYan) [inline repositories/users-repository.js]: Here (in lines 34-37) we are reading the entire file, adding a new user, and completely rewriting the file. It's not effective in terms of RAM usage and time effeciency, especially when we will have a lot of users registered.

File: db/users.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+[
```
Comment (MazanYan) [inline db/users.json:1]: Here we are basically commiting a file with personal data and making it public. Shouldn't it be added to `.gitignore`? And will the application recreate the file if it gets deleted?

File: index.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+app.use((err, req, res, next) => {
+  const status = err.status || HttpCodes.INTERNAL_SERVER_ERROR;
+
+  return res.status(status).json({
+    message: err.message,
+  });
+});
+
+process.on('unhandledRejection', (reason, promise) => {
```
Comment (MazanYan) [inline index.js:38]: I didn't know how to handle this promise rejection error before. It was interesting to know.

File: helpers/is-unique.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+const { getAllUsers } = require('../repositories/users-repository');
+const path = require('path');
+
+const usersPath = path.join(__dirname, '..', 'db', 'users.json');
+
+const isUniqueUser = async email => {
+  try {
+    const allUsers = await getAllUsers(usersPath);
+    const isUnique = allUsers.find(user => user.email === email);
```
Comment (MazanYan) [inline helpers/is-unique.js]: This operation has a linear time complexity what is not effective. I'd suggest to use the Front-end normalization on the allUsers variable and data structure.
Changing from this:
`const allUsers = [
  {
    email: 'kakashi-hatake@mail.com',
    ...
  },
  ...
];`

to this: 
`const allUsers = {
  'kakashi-hatake@mail.com': {
    ...
  },
  'other-user@mail.com': {
    ...
  },
  ...
};`

Then we can find the user using `allUsers[email]` only.

File: package.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  "name": "project-ses-bit-web-api",
+  "version": "1.0.0",
+  "description": "Test task for SES.",
+  "main": "index.js",
+  "scripts": {
+    "start:dev": "nodemon index.js"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Larisa-Antsifrova/project-ses-bit-web-api.git"
```
Comment (MazanYan) [inline package.json]: The repository URL is different now

File: package.json [hunk 6]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "project-ses-bit-web-api",
+  "version": "1.0.0",
+  "description": "Test task for SES.",
+  "main": "index.js",
+  "scripts": {
+    "start:dev": "nodemon index.js"
```
Comment (MazanYan) [inline package.json:8]: Will we have `start:debug` or `start:prod` scripts? If not, we can rename it to `start` only

File: services/coinlayer-api.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
+require('dotenv').config();
+const axios = require('axios');
+
+const COINLAYER_API_KEY = process.env.COINLAYER_API_KEY;
```
Comment (MazanYan) [inline services/coinlayer-api.js]: Why can't we use `const { COINLAYER_API_KEY } = process.env;`? It takes less code space and does not lower readability.

File: controllers/users-controllers.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
+require('dotenv').config();
+const User = require('../repositories/users-repository');
+const isUniqueUser = require('../helpers/is-unique');
+const { HttpCodes } = require('../helpers/constants');
+
+const createUser = async (req, res, next) => {
```
Comment (MazanYan) [inline controllers/users-controllers.js:4]: The controller does too much. In MVC, for example the controller only transfers the user input to model (any UserService etc).

File: controllers/users-controllers.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+    const isUnique = await isUniqueUser(email);
+
+    if (!isUnique) {
+      return res
+        .status(HttpCodes.CONFLICT)
+        .json({ message: 'This email is already in use.' });
+    }
+
+    const hashedPassword = await bcrypt.hash(password, 8);
```
Comment (MazanYan) [inline controllers/users-controllers.js]: Why isn't password encryption a separate service?

File: controllers/users-controllers.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+    if (!isPasswordCorrect) {
+      return res
+        .status(HttpCodes.UNAUTHORIZED)
+        .json({ message: 'Invalid credentials.' });
+    }
+
+    const payload = { id: user.id, name: user.name };
+    const JWT_SECRET_KEY = process.env.JWT_SECRET_KEY;
+    const token = jwt.sign(payload, JWT_SECRET_KEY, { expiresIn: '4h' }); // access token is given for 4 hours
```
Comment (MazanYan) [inline controllers/users-controllers.js]: `expiresIn` clearly explains that the token expires in 4 hours. I think the comment is redundant.

File: controllers/users-controllers.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+    const user = await User.getUserByEmail(email);
+
+    if (!user) {
+      return res
+        .status(HttpCodes.UNAUTHORIZED)
+        .json({ message: 'Invalid credentials.' });
+    }
+
+    const isPasswordCorrect = await bcrypt.compare(password, user.password);
```
Comment (MazanYan) [inline controllers/users-controllers.js]: I suggest to move password encryption and comparing to the separate service.

File: .example.env [hunk 12]
```diff
@@ -0,0 +1,5 @@
+PORT={port number}
+
+JWT_SECRET_KEY={custom secret key for JWT sign}
+
+COINLAYER_API_KEY={api key for bitcoin rates api}
```
Comment (MazanYan) [inline .example.env]: EOF line is missing.

**Review by MazanYan — CHANGES_REQUESTED** [review]
Thank you for a good pull request! I have a few suggestions in comments.

File: controllers/users-controllers.js (hunk 9 above)
Comment (Hellrasier) [reply controllers/users-controllers.js]: Password encryption is better to go to middlewares section

File: controllers/users-controllers.js [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+  } catch (error) {
+    next(error);
+  }
+};
+
+const loginUser = async (req, res, next) => {
+  try {
+    const { email, password } = req.body;
+
+    const user = await User.getUserByEmail(email);
```
Comment (Hellrasier) [inline controllers/users-controllers.js]: Storage is named as User, maybe it is better to create first more abstract "storage" than specify it only for user

File: helpers/api-limiter-config.js [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+const apiLimiterConfig = {
+  windowMs: 900000, // period of 15 minutes
+  max: 100,
+  handler: (req, res, next) => {
+    return res.status(HttpCodes.TOO_MANY_REQUESTS).json({
+      message: 'Too many requrests made. Please try again later.',
+    });
+  },
+};
+
```
Comment (Hellrasier) [inline helpers/api-limiter-config.js:12]: anti-spam and anti-dos system, gj

File: services/coinlayer-api.js [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+        access_key: COINLAYER_API_KEY,
+        target: 'UAH',
+        symbols: 'BTC',
+      },
+    });
+
+    return { timestamp, target, rates };
+  } catch (error) {
+    console.log('Error in fetchUahToBtcRate: ', error.message);
+  }
```
Comment (Hellrasier) [inline services/coinlayer-api.js]: Function returns undefined on error, maybe return error to properly handle it?

File: .example.env (hunk 12 above)
Comment (Larisa-Antsifrova) [reply .example.env]: I get it that EOF stands for End Of File. But I am not sure I get what exactly is expected. Could you please hint me with an example? :)

File: controllers/users-controllers.js (hunk 11 above)
Comment (Larisa-Antsifrova) [reply controllers/users-controllers.js]: Done :)

File: controllers/users-controllers.js (hunk 9 above)
Comment (Larisa-Antsifrova) [reply controllers/users-controllers.js]: That's a great question. Thank you! Moved password hashing and comparing to a separate service in the next commits.

File: controllers/users-controllers.js (hunk 10 above)
Comment (Larisa-Antsifrova) [reply controllers/users-controllers.js]: Noted :) The comment is deleted and the whole JWT generation is moved to a separate service as well.

File: services/coinlayer-api.js (hunk 7 above)
Comment (Larisa-Antsifrova) [reply services/coinlayer-api.js]: Good point! Thanks :) Destructured the variables in other similar places in code as well.

File: package.json (hunk 6 above)
Comment (Larisa-Antsifrova) [reply package.json:8]: I got your point! I have added the _start_ script separately, the one that will be used for deployment, and left _start:dev_ for development.

File: package.json (hunk 5 above)
Comment (Larisa-Antsifrova) [reply package.json]: Haha! V - vnimatelnost. Thank you, Yan! Fixed that :)

File: controllers/users-controllers.js (hunk 8 above)
Comment (Larisa-Antsifrova) [reply controllers/users-controllers.js:4]: Alright, this one took me a while. I've added UserService. But it does not feel to improve anything in particular. If before, the controllers dealt with all the services, now UserService depends on Password and TokenService. On the other hand, I would not like to make my UserService return status codes and response message.
I might not see the long-term benefit here. I would appreciate it, if you could explain or show example :)

File: helpers/is-unique.js (hunk 4 above)
Comment (Larisa-Antsifrova) [reply helpers/is-unique.js]: Nice, that's the first time I've heard of Front-end normalization (I've read this article and I guess that's what you meant: https://alfredayibonte.medium.com/front-end-normalization-8e3d0df7cb47). So, I will ask before action:
Do you suggest it's more efficient to change the structure of data stored in the "database" file from an array of objects to one big object containing all the users by their e-mails as their keys?

File: db/users.json (hunk 2 above)
Comment (Larisa-Antsifrova) [reply db/users.json:1]: That's true :)
In the scope of a test task, I left the DB folder and its file to stay in the project structure. 
So far, no. The DB file is expected to exist)

File: repositories/users-repository.js (hunk 1 above)
Comment (Larisa-Antsifrova) [reply repositories/users-repository.js]: I guess further actions on this one would depend on your suggestion to the thread below :)

**Comment by Larisa-Antsifrova** [comment]
Hi!
Thank you for the suggestions! I've done my best to address most of them:
- Moved password manipulations and token generation to their separate services
- Destructured env variables where needed
- Edited package.json file
- Removed comments
- Created separate UserService to remove some of the logic from controllers

I have resolved the comments that covered the issues I refactored (I hope that's okay). And there are mainly 3 issues left:
- EOF
- Structure of the data stored in DB file
- Controllers taking too much responsibility :)

File: helpers/is-unique.js (hunk 4 above)
Comment (MazanYan) [reply helpers/is-unique.js]: Yes, I meant this normalization. I completed this task by storing all users in the corresponding directory. The `users` directory contained the separate files with users' e-mails as names (user entity primary key). I'm not sure if `fs.readFile(name, callback)` finds the necessary file in constant time but I'm sure it's a more straighforward approach.

File: .example.env (hunk 12 above)
Comment (MazanYan) [reply .example.env]: It's usual for files with code to end with a blank line. I think it serves the same way as a comma in the vertical array/function arguments like this:
```
const longArray = [
  '1',
  '2,235',
  '-1F',
];
```

It makes adding new records and lines of code more easy and makes the file structure more consistent (like with comma example - all records have a comma after them). Linters usually have this rule warning the programmer about missing line in the end of file. GitHub also shows the warning in the pull request.

File: helpers/is-unique.js (hunk 4 above)
Comment (Larisa-Antsifrova) [reply helpers/is-unique.js]: Aha, okay, I see, so that also helped to avoid re-writing the whole file.

File: .example.env (hunk 12 above)
Comment (Larisa-Antsifrova) [reply .example.env]: Wow :) Yan, thanks! That's a  new insight for me :)

>  Linters usually have this rule warning the programmer about missing line in the end of file. GitHub also shows the warning in the pull request.

**Review by MazanYan — APPROVED** [review]

File: .example.env (hunk 12 above)
Comment (Larisa-Antsifrova) [reply .example.env]: Done )

**Review by Hellrasier — APPROVED** [review]

**Pull request merged**